README.md

```diff
@@ -1 +1,141 @@
+All non-functions should have a noun as a name, not a verb.
+
+#### Booleans
+Should start with `is`, `has`, `will` or `should`. Like `isValid` or `hasValues`.
+
+#### Always Affirmative
+Avoid negations. _“Don’t ever not avoid negative logic”_. Prefer `isShown` over `isHidden` or `isEnabled` over `isDisabled`. Do not use names like `notEditable`.
+
+#### Casing
+
+##### Classes, Interfaces & Types
+Are written in Pascal Case naming (start with a capital, the rest camelcase), no underscores, no dashes.
+
+##### Functions, properties, arguments and variables
+Are all written in lower camelCase. 
+
+##### Globally used constants
+Are written in CAPS, words divided using and underscore. Like `CAMPAIGN_ID`.
+
+## Coding
+Every function or class should do **one thing** (and do it good). If it needs to do more than one thing, split it up. Keep your files, classes and functions small. It’s okay to have a file with just a single line.
+
+#### Pure functions
+Prefer writing pure functions, which means they do not manipulate the input arguments or reference/manipulate global state. This makes your code better scalable and testable.
+
+#### Separate Logic From Configuration
+Write code that is reusable, scalable and testable.
+
+#### Do not repeat yourself (DRY)
+* Do not copy code to another place.
+* Avoid using the same string twice in a project.
+* Move shared logic to a shared place.
+* Make sure you do not have to adapt changes on multiple places.
+
+#### Do not use Magic Numbers
+See https://en.wikipedia.org/wiki/Magic_number_(programming)
+
+## Formatting
+All code within a project should have the same formatting. To enforce that we use [Prettier](https://prettier.io/).
+
+## Comments
+#### Documentation comments
+Reconsider if your code really needs comments. Code should be self explaining. Code that needs comments is probably bad.
+
+#### Regular Expressions
+Since regular expressions can be hard to read, they should have a comment that indicate what they do. Especially when they are complex.
+
+#### Commented out code
+Don't leave commented out code into project. You can always find it back in the version control system. If for some reason you want to keep commented out code in the project, add a comment explaining why it is commented out.
+
+#### TODO
+If something needs to be changed or refactored later, add a `// TODO ` comment to indicate what the issue is.
+
+## TypeScript
+Use TypeScript in strict mode and do not allow native JavaScript. Therefore everything must be typed. It’s not needed to type something when TypeScript can resolve the type.
+
+#### Access Modifiers
+Keep your code as strict as possible, so keep all functions and properties `private` unless they have to be `protected` or `public`.
+
+#### Readonly
+In order to be as strict as possible, every property should be set to readonly unless it should writable.
+
+#### Arrays
+Always prefer `ReadonlyArray` over a regular `Array` unless it must be possible to modify the Array.
```

Comment:
I see that we're referring to arrays using the `Array` interface. In most popular languages the `T[]` notation is used for arrays, this type notation is also supported in TypeScript. A `ReadonlyArray<T>` can be created by using the `readonly` keyword since [Typescript  3.4 was released (march 2019)](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#improvements-for-readonlyarray-and-readonly-tuples). The `T[]`  notation is also a little shorter 😇.

```typescript
T[] // Mutable array
readonly T[] // Immutable array

// We cannot assign a new value to the variable or push new elements into the array
const myArray: readonly number[] = []; 

// We can assign a new array to the variable, we cannot push new elements into the array
let myArray: readonly number[] = [];

class MyClass {
  // We cannot assign a new value to the variable or push new elements into the array
  private readonly myProperty: readonly number[] = [];

  // We can assign a new array to the variable, we cannot push new elements into the arrayprivate 
  private myProperty: readonly number[] = [];
}
```

Typescript actually interprets the `Array<T>` and `ReadonlyArray<T>` as `T[]` and `readonly T[]`.

![image](https://user-images.githubusercontent.com/5376407/83916019-bac8fb00-a774-11ea-9bc2-efc75228a4ef.png)
![image](https://user-images.githubusercontent.com/5376407/83915975-9bca6900-a774-11ea-8db2-60f450a9ad63.png)

The `[]` notation also allows us to create Tuple types.

![image](https://user-images.githubusercontent.com/5376407/83914594-d088f100-a771-11ea-8564-98ac3e072106.png)

Reply:
I personally don't like the `MyType[]` notation for Array's for 2 reasons:
- Consistency. `Array` is just a generic type, and all (other) generic types are typed using `<MyType>`
- `MyType[]` is easily confused with `[MyType]`, which is also valid TypeScript. But `[MyType]` means that this is an Array (with a length of 1) and where only the first element is typed.

Reply:
So you should use the `[MyType, MyOtherType]` for tuples, not for Array's.

Reply:
I'm a sucker for Consistency, so personally I like to use **generics** everywhere the same:
```ts
let a: Array<string>;
let s: Set<string>;
let o: Record<string, number>;
let m: Map<string, number>;
let t: [string, string]; // the odd one

let a: string[]; // meh
```

`readonly` (type modifier) only works on those two below. If we ever get a ReadonlySet, it might also start woking on that.
```ts
let a: ReadonlyArray<string>;
let t: Readonly<[string, string]>;

// vs
let a: readonly string[];
let t: readonly [string, string];
```

Even though it's nice syntactic sugar for people already using the other shorthands, it can be confusing with `readonly` on class properties, which work as a _modifier_ (changing the binding, not the type), so you would have to use both: `readonly s: readonly string[];`

In general, even though `[]` is sorter compared to `Array<>`, I like the generic for 2 other reasons:
* It's easier to see the difference between the single string, and an array of strings
* It focusses on the important part of the type (it being an `Array`), that there are strings inside is less relevant for what you want to do with it.

Fun fact, in TS before generics, using `[]` was the only way to annotate arrays. So we deliberately switched over to start using generic syntax consistently.

Reply:
Good point. Especially if you should call the Generic type `T` or not.

Reply:
I also wonder if we should add a linting rule that forbids the usages of a one length Tuple (`[MyType]`), since this is barely used and a frequently made mistake if you actually wanted to write a regular Array. Especially in project with GraphQL, since `[MyType]` is how you define an Array in GraphlQL

Reply:
Can't hurt to add it, and remove it later if it turns out there are indeed quite some use cases.

Reply:
I created an issue for this #19